yt_idv/coordinate_utilities.pyx

```diff
@@ -0,0 +1,526 @@
+                           np.float64_t[3] dxyz_1,
+                           np.float64_t[3] xyz,
+                           np.float64_t[3] dxyz) noexcept nogil:
+
+    # find the effective bounding box given 2 others
+    cdef np.float64_t le_i, re_i
+    cdef int idim
+    cdef int ndim = 3
+
+    for idim in range(ndim):
+        le_i = fmin(xyz_0[idim] - dxyz_0[idim]/2.,
+                    xyz_1[idim] - dxyz_1[idim]/2.)
+        re_i = fmax(xyz_0[idim] + dxyz_0[idim]/2.,
+                    xyz_1[idim] + dxyz_1[idim]/2.)
+        xyz[idim] = (le_i + re_i)/2.
+        dxyz[idim] = re_i - le_i
+
+
+cdef class MixedCoordBBox:
+    # abstract class for calculating cartesian bounding boxes
+    # from non-cartesian grid elements.
+    cdef int get_cartesian_bbox(self,
+                                np.float64_t pos0,
+                                np.float64_t pos1,
+                                np.float64_t pos2,
+                                np.float64_t dpos0,
+                                np.float64_t dpos1,
+                                np.float64_t dpos2,
+                                np.float64_t xyz_i[3],
+                                np.float64_t dxyz_i[3]
+                                ) noexcept nogil:
+        pass
+
+
+cdef class SphericalMixedCoordBBox(MixedCoordBBox):
+    # Cartesian bounding boxes of spherical grid elements
+    cdef int get_cartesian_bbox(self,
+                        np.float64_t pos0,  # r
+                        np.float64_t pos1,  # theta
+                        np.float64_t pos2,  # phi
+                        np.float64_t dpos0, # r
+                        np.float64_t dpos1, # theta
+                        np.float64_t dpos2, # phi
+                        np.float64_t[3] xyz_i,
+                        np.float64_t[3] dxyz_i,
+                        ) noexcept nogil:
+        """
+        Calculate the cartesian bounding box for a single spherical volume element.
+
+        Parameters
+        ----------
+        pos0: float
+            radius value at element center
+        pos1: float
+            co-latitude angle value at element center, in range (0, pi) (theta in yt)
+        pos2: float
+            azimuthal angle value at element center, in range (0, 2pi) (phi in yt)
+        dpos0: float
+            radial width of element
+        dpos1: float
+            co-latitude angular width of element
+        pos2: float
+            azimuthal anglular width of element
+        xyz_i:
+            3-element array to store the center of the resulting bounding box
+        dxyz_i:
+            3-element array to store the width of the resulting bounding box
+
+        Returns
+        -------
+        int
+            error code: 0 for success, 1 if something went wrong...
+
+        Note: this function is recursive! If either angular width is above a cutoff (of
+        0.2 radians or about 11 degrees), the element will be recursively divided.
```

Comment:
Why 0.2 rad / 11 deg?

Reply:
I... don't remember. Best guess looking at it now is that when playing with what to use, 10 degrees worked well without adding too much extra computation, but I wanted it to be a nice number in radians?